File: compatibility-tests/sdk-test-go/tests/cloudwatch_test.go
package tests

import (
	"context"
	"testing"
	"time"

	"floci-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/cloudwatch"
	cwtypes "github.com/aws/aws-sdk-go-v2/service/cloudwatch/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestCloudWatch(t *testing.T) {
	ctx := context.Background()
	svc := testutil.CloudWatchClient()
	namespace := "GoTest"

	t.Run("PutMetricData", func(t *testing.T) {
		_, err := svc.PutMetricData(ctx, &cloudwatch.PutMetricDataInput{
			Namespace: aws.String(namespace),
			MetricData: []cwtypes.MetricDatum{
				{
					MetricName: aws.String("RequestCount"),
					Value:      aws.Float64(42),
					Unit:       cwtypes.StandardUnitCount,
					Timestamp:  aws.Time(time.Now()),
				},
			},
		})
		require.NoError(t, err)
	})

	t.Run("ListMetrics", func(t *testing.T) {
		r, err := svc.ListMetrics(ctx, &cloudwatch.ListMetricsInput{Namespace: aws.String(namespace)})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Metrics)
	})

	t.Run("GetMetricStatistics", func(t *testing.T) {
		now := time.Now()
		r, err := svc.GetMetricStatistics(ctx, &cloudwatch.GetMetricStatisticsInput{
			Namespace:  aws.String(namespace),
			MetricName: aws.String("RequestCount"),
			StartTime:  aws.Time(now.Add(-5 * time.Minute)),
			EndTime:    aws.Time(now.Add(time.Minute)),
			Period:     aws.Int32(60),
			Statistics: []cwtypes.Statistic{cwtypes.StatisticSum},
		})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Datapoints)
	})

	t.Run("PutMetricDataWithStatisticValues", func(t *testing.T) {
		now := time.Now()
		namespace := "GoTestStatisticValues"

		// Put metric data with pre-calculated statistics
		_, err := svc.PutMetricData(ctx, &cloudwatch.PutMetricDataInput{
			Namespace: aws.String(namespace),
			MetricData: []cwtypes.MetricDatum{
				{
					MetricName: aws.String("AggregatedMetric"),
					StatisticValues: &cwtypes.StatisticSet{
						SampleCount: aws.Float64(5.0),
						Sum:         aws.Float64(150.0),
						Minimum:     aws.Float64(20.0),
						Maximum:     aws.Float64(40.0),
					},
					Unit: cwtypes.StandardUnitCount,
				},
			},
		})
		require.NoError(t, err)

		// Query back the statistics
		r, err := svc.GetMetricStatistics(ctx, &cloudwatch.GetMetricStatisticsInput{
			Namespace:  aws.String(namespace),
			MetricName: aws.String("AggregatedMetric"),
			StartTime:  aws.Time(now.Add(-time.Hour)),
			EndTime:    aws.Time(now.Add(time.Minute)),
			Period:     aws.Int32(3600),
			Statistics: []cwtypes.Statistic{
				cwtypes.StatisticSum,
				cwtypes.StatisticAverage,
				cwtypes.StatisticMinimum,
				cwtypes.StatisticMaximum,
				cwtypes.StatisticSampleCount,
			},
		})

		require.NoError(t, err)
		assert.NotEmpty(t, r.Datapoints)

		dp := r.Datapoints[0]
		assert.NotNil(t, dp.Sum)
		assert.Equal(t, 150.0, *dp.Sum)
		assert.NotNil(t, dp.SampleCount)
		assert.Equal(t, float64(5), *dp.SampleCount)
		assert.NotNil(t, dp.Minimum)
		assert.Equal(t, 20.0, *dp.Minimum)
		assert.NotNil(t, dp.Maximum)
		assert.Equal(t, 40.0, *dp.Maximum)
		assert.NotNil(t, dp.Average)
		assert.Equal(t, 30.0, *dp.Average) // sum / sampleCount
	})
}

File: compatibility-tests/sdk-test-node/tests/cloudwatch.test.ts
/**
 * CloudWatch Metrics integration tests.
 */

import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import {
  CloudWatchClient,
  PutMetricDataCommand,
  GetMetricStatisticsCommand,
  ListMetricsCommand,
  PutMetricAlarmCommand,
  DescribeAlarmsCommand,
  DeleteAlarmsCommand,
} from '@aws-sdk/client-cloudwatch';
import { makeClient, uniqueName } from './setup';

describe('CloudWatch Metrics', () => {
  let cw: CloudWatchClient;
  let namespace: string;
  let alarmName: string;

  beforeAll(() => {
    cw = makeClient(CloudWatchClient);
    namespace = `Floci/Test/${uniqueName()}`;
    alarmName = `test-alarm-${uniqueName()}`;
  });

  afterAll(async () => {
    try {
      await cw.send(new DeleteAlarmsCommand({ AlarmNames: [alarmName] }));
    } catch {
      // ignore
    }
  });

  it('should put metric data', async () => {
    await cw.send(
      new PutMetricDataCommand({
        Namespace: namespace,
        MetricData: [{ MetricName: 'TestMetric', Value: 42.0, Unit: 'Count' }],
      })
    );
  });

  it('should put metric data batch', async () => {
    await cw.send(
      new PutMetricDataCommand({
        Namespace: namespace,
        MetricData: [
          { MetricName: 'Latency', Value: 100, Unit: 'Milliseconds' },
          { MetricName: 'Errors', Value: 0, Unit: 'Count' },
        ],
      })
    );
  });

  it('should list metrics', async () => {
    const response = await cw.send(new ListMetricsCommand({ Namespace: namespace }));
    expect(response.Metrics?.length).toBeGreaterThan(0);
  });

  it('should get metric statistics', async () => {
    const now = new Date();
    const start = new Date(now.getTime() - 3600000);

    const response = await cw.send(
      new GetMetricStatisticsCommand({
        Namespace: namespace,
        MetricName: 'TestMetric',
        StartTime: start,
        EndTime: now,
        Period: 3600,
        Statistics: ['Sum', 'Average', 'Maximum'],
      })
    );
    expect(response.Datapoints).toBeDefined();
  });

  it('should put metric data with statistic values', async () => {
    const now = new Date();
    const start = new Date(now.getTime() - 3600000);

    // Put metric data with pre-calculated statistics
    await cw.send(
      new PutMetricDataCommand({
        Namespace: namespace,
        MetricData: [
          {
            MetricName: 'AggregatedMetric',
            StatisticValues: {
              SampleCount: 5,
              Sum: 150,
              Minimum: 20,
              Maximum: 40,
            },
            Unit: 'Count',
          },
        ],
      })
    );

    // Query back the statistics
    const response = await cw.send(
      new GetMetricStatisticsCommand({
        Namespace: namespace,
        MetricName: 'AggregatedMetric',
        StartTime: start,
        EndTime: now,
        Period: 3600,
        Statistics: ['Sum', 'Average', 'Minimum', 'Maximum', 'SampleCount'],
      })
    );
    
    expect(response.Datapoints).toBeDefined();
    expect(response.Datapoints?.length).toBeGreaterThan(0);
    
    const dp = response.Datapoints?.[0];
    expect(dp?.Sum).toBe(150);
    expect(dp?.SampleCount).toBe(5);
    expect(dp?.Minimum).toBe(20);
    expect(dp?.Maximum).toBe(40);
    expect(dp?.Average).toBe(30); // sum / sampleCount
  });

  it('should put metric alarm', async () => {
    await cw.send(
      new PutMetricAlarmCommand({
        AlarmName: alarmName,
        MetricName: 'TestMetric',
        Namespace: namespace,
        Statistic: 'Average',
        Period: 60,
        EvaluationPeriods: 1,
        Threshold: 100,
        ComparisonOperator: 'GreaterThanThreshold',
      })
    );
  });

  it('should describe alarms', async () => {
    const response = await cw.send(
      new DescribeAlarmsCommand({ AlarmNames: [alarmName] })
    );
    expect(response.MetricAlarms?.some((a) => a.AlarmName === alarmName)).toBe(true);
  });

  it('should delete alarms', async () => {
    await cw.send(new DeleteAlarmsCommand({ AlarmNames: [alarmName] }));
    alarmName = '';
  });
});
